### PR TITLE
Gitlab 274: Attachment security

### PR DESCRIPTION
**Pull request by quynhn**
Do not review please

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/AttachmentManager.java [hunk 1]
```diff
@@ -19,14 +19,12 @@
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/AttachmentManager.java]: Please keep these methods ;-)

Even if you think they are not useful, they are:
 - **getOwnerMessageIds** is re-inserted next commit.
 - **storeAttachmentsForMessage** is the method that SHOULD be used. (I even wonder if *storeAttachmentsForMessage* alone makes sens)

File: server/container/cli/src/main/java/org/apache/james/cli/probe/impl/JmxMailboxProbe.java [hunk 2]
```diff
@@ -30,11 +30,15 @@
...
 import org.apache.james.adapter.mailbox.ReIndexerManagementMBean;
 import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.AttachmentId;
 import org.apache.james.mailbox.model.ComposedMessageId;
 import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.mailbox.model.MessageId;
 import org.apache.james.mailbox.store.mail.model.Mailbox;
 import org.apache.james.mailbox.store.probe.MailboxProbe;
 
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
Comment (chibenwa) [inline server/container/cli/src/main/java/org/apache/james/cli/probe/impl/JmxMailboxProbe.java]: Not the good one.

File: server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/cucumber/MemoryDownloadCucumberTest.java [hunk 3]
```diff
@@ -25,7 +25,8 @@
 import cucumber.api.junit.Cucumber;
 
 @RunWith(Cucumber.class)
-@CucumberOptions(features={"classpath:cucumber/DownloadEndpoint.feature", "classpath:cucumber/DownloadGet.feature", "classpath:cucumber/DownloadPost.feature"},
+@CucumberOptions(features={"classpath:cucumber/DownloadGet.feature"},
+//    @CucumberOptions(features={"classpath:cucumber/DownloadEndpoint.feature", "classpath:cucumber/DownloadGet.feature", "classpath:cucumber/DownloadPost.feature"},
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/cucumber/MemoryDownloadCucumberTest.java]: Oups

File: server/protocols/jmap-integration-testing/pom.xml [hunk 4]
```diff
@@ -34,6 +34,8 @@
 
     <modules>
         <module>jmap-integration-testing-common</module>
+        <module>cassandra-jmap-integration-testing</module>
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/pom.xml]: Oups

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/DownloadServlet.java [hunk 5]
```diff
@@ -100,15 +112,23 @@ private void respondAttachmentAccessToken(MailboxSession mailboxSession, Downloa
         }
     }
 
-    private boolean attachmentExists(MailboxSession mailboxSession, String blobId) throws MailboxException {
+    private Optional<Blob> getBlob(MailboxSession mailboxSession, String blobId) throws MailboxException {
         try {
-            blobManager.retrieve(BlobId.fromString(blobId), mailboxSession);
-            return true;
+            Collection<MessageId> ownerMessageIds = attachmentManager.getOwnerMessageId(AttachmentId.from(blobId), mailboxSession);
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/DownloadServlet.java]: Security SHOULD be enforced by the AttachmentManager.

The AttachmentManager should not return attachments I do not have the rights on.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/probe/MailboxProbe.java [hunk 6]
```diff
@@ -54,4 +56,6 @@ ComposedMessageId appendMessage(String username, MailboxPath mailboxPath, InputS
     void reIndexAll() throws Exception;
 
     Collection<String> listSubscriptions(String user) throws Exception;
+
+    Collection<MessageId> getOwnerMessageId(AttachmentId attachmentId, String user) throws Exception;
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/probe/MailboxProbe.java]: See below, but you don't need this as you are doing a testing anti-pattern.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 7]
```diff
@@ -387,6 +393,12 @@ public void assertContentLength(int size) throws IOException {
         assertThat(response.getFirstHeader("Content-Length").getValue()).isEqualTo(String.valueOf(size));
     }
 
+    @Then("^user \"([^\"]*)\" request the attachment \"([^\"]*)\" is still there$")
+    public void assertAttachment(String username, String attachmentId) throws Exception {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: I don't care about these lower level details. What I care here is:

 - Can user A downloads user B attacments/messages ? The answer is no.

I don't care about the "how" what is an implementation detail, and SHOULD be abstracted from your tests

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadGet.feature [hunk 8]
```diff
@@ -36,9 +36,16 @@ Feature: Download GET
...
+    Scenario: Getting a message then getting its blob
     Given the user has a message "m1" in "INBOX" mailbox with subject "my test subject", content "testmail"
     And the user ask for messages "m1"
     When "username@domain.tld" downloads the message by its blobId
     Then the user should receive that blob
-    And the blob size is 36
\ No newline at end of file
+    And the blob size is 36
+
+    Scenario: Getting an attachment should be security
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadGet.feature]: Indent.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/probe/MailboxProbe.java [hunk 9]
```diff
@@ -54,4 +56,6 @@ ComposedMessageId appendMessage(String username, MailboxPath mailboxPath, InputS
     void reIndexAll() throws Exception;
 
     Collection<String> listSubscriptions(String user) throws Exception;
+
+    Collection<MessageId> getOwnerMessageIds(AttachmentId attachmentId, String user) throws Exception;
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/probe/MailboxProbe.java]: Why it's in MailboxProbe ? It doesn't seem related to a mailbox

**Comment by mbaechler** [comment]
my work is here : https://github.com/linagora/james-project/pull/977
